File: leetcode/n-queensII.py
# Follow up for N-Queens problem.
# Now, instead outputting board configurations, return the total number of 
# distinct solutions.

class Solution:
     # @param {integer} n
     # @return {integer}
     def totalNQueens(self, n):
        mx, d1, d2 = [False] * n, [False] * (2 * n), [False] * (2 * n)
        stack, solution, result = [(0, 0, False)], [0] * n, 0
        while stack:
            y, x, flag = stack.pop()
            if y == n and flag == False:
                result += 1
            elif not flag:
                if x + 1 < n: stack.append((y, x + 1, False))
                if not mx[x] and not d1[x + y]  and not d2[n - 1 + x - y]:
                    mx[x] = d1[x + y] = d2[n - 1 + x - y] = True
                    solution[y] = x
                    stack.append((y, x, True))
                    stack.append((y + 1, 0, False))
            else:
                mx[x] = d1[x + y] = d2[n - 1 + x - y] = False
        return result

if __name__ == "__main__":
    solution = Solution()
    print solution.totalNQueens(12)

File: leetcode/n-queensII.js
/* Follow up for N-Queens problem.
 * Now, instead outputting board configurations, return the total number of 
 * distinct solutions.
 */

/**
 * @param {number} n
 * @return {number}
 */
var totalNQueens = function(n) {
    var mx = new Array(n), d1 = new Array(2 * n), d2 = new Array(2 * n);
    mx.fill(false);
    d1.fill(false);
    d2.fill(false);
    var solution = new Array(n), temp = new Array(n), result = 0;
    solution.fill(0);
    for (var stack = [[0, 0, false]]; stack.length > 0; ){
        var entry = stack.pop();
        var y = entry[0];
        var x = entry[1];
        if (entry[0] === n) {
            ++result;
        } else if (!entry[2]) {
            if (x + 1 < n)
                stack.push([y, x + 1, false]);
            if (!mx[x] && !d1[x + y] && !d2[n - 1 + x - y]) {
                mx[x] = d1[x + y] = d2[n - 1 + x - y] = true;
                solution[y] = x;
                stack.push([y, x, true], [y + 1, 0, false]);
            }
        } else
            mx[x] = d1[x + y] = d2[n - 1 + x - y] = false;
    }
    return result;
};
